README.md
## Sneeker mod for Minetest

---
### Description:

An explosive nuisance for [Minetest](http://minetest.net/).

![screenshot](screenshot.png)

---
### Licensing:

- Code: [MIT](LICENSE.txt)
  - Original by Rui: WTFPL
- tnt_function code: [MIT](tnt_function.lua)

---
### Usage:

Settings:
- ***sneeker.lifetime***
  - How long (in seconds) sneeker remains in world after spawn.
  - type: int
  - default: 900 (15 minutes)
- ***sneeker.boom_gain***
  - Loudness of explosion.
  - type: float
  - default: 1.5
- ***sneeker.spawn_require_player_nearby***
  - Determines whether or not a player must be close for spawn to occur.
  - type: bool
  - default: true
- ***sneeker.spawn_player_radius***
  - Distance a player must be within for spawn to occur.
  - type: int
  - default: 100
- ***sneeker.despawn_player_far***
  - If enabled, mobs not near any players will despawn.
  - type: bool
  - default: true
- ***sneeker.despawn_player_radius***
  - Distance determining if a player is near enough to prevent despawn.
  - type: int
  - default: 500
- ***sneeker.spawn_chance***
  - Sets possibility for spawn.
  - type: int
  - default: 10000
- ***sneeker.spawn_interval***
  - Sets frequency of spawn chance.
  - type: int
  - default: 240 (4 minutes)
- ***sneeker.spawn_minlight***
  - Sets the minimum light that a node must have for spawn to occur.
  - type: int
  - default: 0
- ***sneeker.spawn_maxlight***
  - Sets the maximum light that a node can have for spawn to occur.
  - type: int
  - default: 4
- ***sneeker.spawn_minheight***
  - Sets the maximum light that a node can have for spawn to occur.
  - type: int
  - default: -31000
- ***sneeker.spawn_maxheight***
  - Sets the lowest position at which sneeker can spawn.
  - type: int
  - default 31000
- ***sneeker.spawn_mapblock_limit***
  - Limits the number of entities that can spawn per mapblock (16x16x16).
  - type: int
  - default: 1

---
### Links:

- [Forum thread](https://forum.minetest.net/viewtopic.php?t=26685)
- [Git repo](https://github.com/AntumMT/mod-sneeker)
- [Original forum thread](https://forum.minetest.net/viewtopic.php?t=11891)


CHANGES.txt

1.0
- Forked & renamed from Rui's original "creeper" mod.
- Updated for Minetest 5.x API.
- Changed to MIT license.
- Changed spawn nodes:
  - default:dirt_with_dry_grass
  - default:dry_dirt
  - default:dry_dirt_with_dry_grass
  - default:desert_sand
- Spawns on "nether:rack" if nether mod available.


init.lua

sneeker = {}
sneeker.modname = core.get_current_modname()
sneeker.modpath = core.get_modpath(sneeker.modname)

local debugging = core.settings:get_bool("enable_debug_mods", false)
sneeker.log = function(lvl, msg)
	if lvl == "debug" and not debugging then return end

	if msg == nil then
		msg = lvl
		lvl = nil
	end

	msg = "[" .. sneeker.modname .. "] " .. msg
	if lvl == "debug" then
		msg = "[DEBUG]" .. msg
		lvl = nil;
	end

	if not lvl then
		core.log(msg)
	else
		core.log(lvl, msg)
	end
end

local scripts = {
	"settings",
	"tnt_function",
	"spawn",
}

for _, script in ipairs(scripts) do
	dofile(sneeker.modpath .. "/" .. script .. ".lua")
end


local function jump(self, pos, direction)
	local velocity = self.object:get_velocity()
	if core.registered_nodes[core.get_node(pos).name].climbable then
		self.object:set_velocity({x=velocity.x, y=4, z=velocity.z})
		return
	end

	local spos = {x=pos.x+direction.x, y=pos.y, z=pos.z+direction.z}
	local node = core.get_node_or_nil(spos)
	spos.y = spos.y+1
	local node2 = core.get_node_or_nil(spos)
	local def, def2 = {}
	if node and node.name then
		def = core.registered_items[node.name]
	end
	if node2 and node2.name then
		def2 = core.registered_items[node2.name]
	end
	if def and def.walkable
	and def2 and not def2.walkable
	and def.drawtype ~= "fencelike" then
		self.object:set_velocity({
			x = velocity.x*2.2,
			y = self.jump_height,
			z = velocity.z*2.2
		})
	end
end

local function random_turn(self)
	if self.turn_timer > math.random(2, 5) then
		local select_turn = math.random(1, 3)
		if select_turn == 1 then
			self.turn = "left"
		elseif select_turn == 2 then
			self.turn = "right"
		elseif select_turn == 3 then
			self.turn = "straight"
		end
		self.turn_timer = 0
		self.turn_speed = 0.05*math.random()
	end
end

local def = {
	hp_max = 20,
	physical = true,
	collisionbox = {-0.25, -0.7, -0.25, 0.25, 0.8, 0.25},
	visual = "mesh",
	mesh = "character.b3d",
	textures = {"sneeker.png"},
	makes_footstep_sound = false,

	-- Original
	animation = {
		stand_START = 0,
		stand_END = 79,
		walk_START = 168,
		walk_END = 187
	},
	walk_speed = 1.5,
	jump_height = 5,
	animation_speed = 30,
	knockback_level = 2
}

def.on_activate = function(self, staticdata)
	self.yaw = 0
	self.anim = 1
	self.timer = 0
	self.visualx = 1
	self.jump_timer = 0
	self.turn_timer = 0
	self.turn_speed = 0
	self.powered = false
	self.knockback = false
	self.state = math.random(1, 2)
	self.old_y = self.object:get_pos().y
	self.lifetime = sneeker.lifetime
	self.lifetimer = 0

	local data = core.deserialize(staticdata)
	if data and type(data) == "table" then
		if data.powered == true then
			self.powered = true
			self.object:set_properties({textures = {"sneeker_powered.png"}})
		end
	else
		if math.random(0, 20) == 20 then
			self.powered = true
			self.object:set_properties({textures = {"sneeker_powered.png"}})
		end
	end
end


local function isnan(n)
	return tostring(n) == tostring((-1)^.5)
end

def.on_step = function(self, dtime)
	-- update lifetime timer
	-- FIXME: this is longer than realtime
	self.lifetimer = self.lifetimer + dtime
	if self.lifetimer >= self.lifetime then
		-- TODO: should have a death animation
		self.object:remove()
		return true
	end

	if self.knockback then
		return false
	end

	local ANIM_STAND = 1
	local ANIM_WALK  = 2

	local pos = self.object:get_pos()
	local yaw = self.object:get_yaw()
	local inside = core.get_objects_inside_radius(pos, 10)
	local walk_speed = self.walk_speed
	local animation = self.animation
	local anim_speed = self.animation_speed
	local velocity = self.object:get_velocity()

	self.timer = self.timer+0.01
	self.turn_timer = self.turn_timer+0.01
	self.jump_timer = self.jump_timer+0.01

	if not self.chase
	and self.timer > math.random(2, 5) then
		if math.random() > 0.8 then
			self.state = "stand"
		else
			self.state = "walk"
		end
		self.timer = 0
	end

	if self.turn == "right" then
		self.yaw = self.yaw+self.turn_speed
		self.object:set_yaw(self.yaw)
	elseif self.turn == "left" then
		self.yaw = self.yaw-self.turn_speed
		self.object:set_yaw(self.yaw)
	end

	if self.chase and self.visualx < 2 then
		if self.hiss == false then
			core.sound_play("sneeker_hiss", {pos=pos, gain=1.5, max_hear_distance=2*64})
		end
		self.visualx = self.visualx+0.05
		self.object:set_properties({
			visual_size = {x=self.visualx, y=1}
		})
		self.hiss = true
	elseif self.visualx > 1 then
		self.visualx = self.visualx-0.05
		self.object:set_properties({
			visual_size = {x=self.visualx, y=1}
		})
		self.hiss = false
	end

	self.chase = false

	for  _, object in ipairs(inside) do
		if object:is_player() then
			self.state = "chase"
		end
	end

	if self.state == "stand" then
		if self.anim ~= ANIM_STAND then
			self.object:set_animation({x=animation.stand_START, y=animation.stand_END}, anim_speed, 0)
			self.anim = ANIM_STAND
		end

		random_turn(self)

		if velocity.x ~= 0
		or velocity.z ~= 0 then
			self.object:set_velocity({x=0, y=velocity.y, z=0})
		end
	end

	if self.state == "walk" then
		if self.anim ~= ANIM_WALK then
			self.object:set_animation({x=animation.walk_START, y=animation.walk_END}, anim_speed, 0)
			self.anim = ANIM_WALK
		end

		self.direction = {x=math.sin(yaw)*-1, y=-10, z=math.cos(yaw)}
		if self.direction then
			self.object:set_velocity({x=self.direction.x*walk_speed, y=velocity.y, z=self.direction.z*walk_speed})
		end

		random_turn(self)

		local velocity = self.object:get_velocity()

		if self.turn_timer > 1 then
			local direction = self.direction
			local npos = {x=pos.x+direction.x, y=pos.y+0.2, z=pos.z+direction.z}
			if velocity.x == 0 or velocity.z == 0
			or core.registered_nodes[core.get_node(npos).name].walkable then
				local select_turn = math.random(1, 2)
				if select_turn == 1 then
					self.turn = "left"
				elseif select_turn == 2 then
					self.turn = "right"
				end
				self.turn_timer = 0
				self.turn_speed = 0.05*math.random()
			end
		end

		-- Jump
		if self.jump_timer > 0.2 then
			jump(self, pos, self.direction)
		end
	end

	if self.state == "chase" then
		if self.anim ~= ANIM_WALK then
			self.object:set_animation({x=animation.walk_START, y=animation.walk_END}, anim_speed, 0)
			self.anim = ANIM_WALK
		end

		self.turn = "straight"

		local inside_2 = core.get_objects_inside_radius(pos, 2)

		-- Boom
		if #inside_2 ~= 0 then
			for  _, object in ipairs(inside_2) do
				if object:is_player() and object:get_hp() ~= 0 then
					self.chase = true
					if self.visualx >= 2 then
						self.object:remove()
						sneeker.boom(pos, self.powered)
						core.sound_play("sneeker_explode", {pos=pos, gain=sneeker.boom_gain, max_hear_distance=2*64})
						return true
					end
				end
			end
		end

		if #inside ~= 0 then
			for  _, object in ipairs(inside) do
				if object:is_player() and object:get_hp() ~= 0 then
					if #inside_2 ~= 0 then
						for _, object in ipairs(inside_2) do
							-- Stop move
							if object:is_player() then
								if self.anim ~= ANIM_STAND then
									self.object:set_animation({x=animation.stand_START, y=animation.stand_END}, anim_speed, 0)
									self.anim = ANIM_STAND
								end
								self.object:set_velocity({x=0, y=velocity.y, z=0})
								return
							end
						end
					end

					local ppos = object:get_pos()
					self.vec = {x=ppos.x-pos.x, y=ppos.y-pos.y, z=ppos.z-pos.z}
					self.yaw = math.atan(self.vec.z/self.vec.x)+math.pi^2
					if ppos.x > pos.x then
						self.yaw = self.yaw+math.pi
					end
					self.yaw = self.yaw-2
					self.object:set_yaw(self.yaw)
					self.direction = {x=math.sin(self.yaw)*-1, y=0, z=math.cos(self.yaw)}

					local direction = self.direction

					-- FIXME: hack
					local can_set = true
					for _, c in ipairs({direction.x*2.5, direction.z*2.5}) do
						if isnan(c) then
							can_set = false
							break
						end
					end

					if can_set then
						self.object:set_velocity({x=direction.x*2.5, y=velocity.y, z=direction.z*2.5})
					end

					-- Jump
					if self.jump_timer > 0.2 then
						jump(self, pos, direction)
					end
				end
			end
		else
			self.state = "stand"
		end
	end

	-- Swim
	local node = core.get_node(pos)
	if core.get_item_group(node.name, "water") ~= 0 then
		self.object:set_acceleration({x=0, y=1, z=0})
		local velocity = self.object:get_velocity()
		if self.object:get_velocity().y > 5 then
			self.object:set_velocity({x=0, y=velocity.y-velocity.y/2, z=0})
		else
			self.object:set_velocity({x=0, y=velocity.y+1, z=0})
		end
	else
		self.object:set_acceleration({x=0, y=-10, z=0})
	end

	return true
end

def.on_punch = function(self, puncher, time_from_last_punch, tool_capabilities, dir)
	if self.knockback == false then
		local knockback_level = self.knockback_level
		self.object:set_velocity({x=dir.x*knockback_level, y=3, z=dir.z*knockback_level})
		self.knockback = true
		core.after(0.6, function()
			self.knockback = false
		end)
	end
	if self.object:get_hp() < 1 then
		local pos = self.object:get_pos()
		local x = 1/math.random(1, 5)*dir.x
		local z = 1/math.random(1, 5)*dir.z
		local p = {x=pos.x+x, y=pos.y, z=pos.z+z}
		local node = core.get_node_or_nil(p)
		if node == nil or not node.name or node.name ~= "air" then
			p = pos
		end
		local obj = core.add_item(p, {name="tnt:gunpowder", count=math.random(0, 2)})
	end
end

def.get_staticdata = function(self)
	return core.serialize({
		powered = self.powered
	})
end

core.register_entity("sneeker:sneeker", def)

core.register_craftitem("sneeker:spawnegg", {
	description = "Sneeker Spawn Egg",
	inventory_image = "sneeker_spawnegg.png",
	stack_max = 64,
	on_place = function(itemstack, placer, pointed_thing)
		if pointed_thing.type == "node" then
			local pos = pointed_thing.above
			pos.y = pos.y+1
			core.add_entity(pos, "sneeker:sneeker")
			if not core.settings:get_bool("creative_mode", false) then
				itemstack:take_item()
			end
			return itemstack
		end
	end
})


spawn.lua

local spawn_nodes = {
	"default:dirt_with_dry_grass",
	"default:dry_dirt",
	"default:dry_dirt_with_dry_grass",
	"default:desert_sand",
}

if core.global_exists("nether") then
	table.insert(spawn_nodes, "nether:rack")
end

for _, node_name in ipairs(spawn_nodes) do
	if not core.registered_nodes[node_name] then
		sneeker.log("warning", "Invalid node for spawn: " .. node_name)
	end
end


core.register_abm({
	nodenames = spawn_nodes,
	neighbors = {"air"},
	interval = sneeker.spawn_interval,
	chance = sneeker.spawn_chance,
	action = function(pos, node, aoc, aocw)
		if aoc >= sneeker.spawn_mapblock_limit then return end

		-- check above target node
		pos.y = pos.y+1

		if pos.y > sneeker.spawn_maxheight or pos.y < sneeker.spawn_minheight then
			return
		end

		-- needs two vertical air nodes to spawn
		for _, node_pos in ipairs({pos, {x=pos.x, y=pos.y+1, z=pos.z}}) do
			if core.get_node(node_pos).name ~= "air" then
				return
			end
		end

		local llevel = core.get_node_light(pos)
		if not llevel or llevel > sneeker.spawn_maxlight or llevel < sneeker.spawn_minlight then
			return
		end

		local spawned = core.add_entity(pos, "sneeker:sneeker")
		if not spawned then
			sneeker.log("warning", "Failed to spawn at: "
				.. tostring(math.floor(pos.x))
				.. "," .. tostring(math.floor(pos.y))
				.. "," .. tostring(math.floor(pos.z)))
		else
			sneeker.log("debug", "Spawned at: "
				.. tostring(math.floor(pos.x))
				.. "," .. tostring(math.floor(pos.y))
				.. "," .. tostring(math.floor(pos.z)))
		end
	end
})


tnt_function.lua
-- From TNT

local cid_data = {}
local radius = tonumber(core.settings:get("tnt_radius") or 3)
local large_radius = 5
local loss_prob = {
	["default:cobble"] = 3,
	["default:dirt"] = 4,
}
core.after(0, function()
	for name, def in pairs(core.registered_nodes) do
		cid_data[core.get_content_id(name)] = {
			name = name,
			drops = def.drops,
			flammable = def.groups.flammable,
		}
	end
end)

local function rand_pos(center, pos, radius)
	pos.x = center.x + math.random(-radius, radius)
	pos.z = center.z + math.random(-radius, radius)
end

local function eject_drops(drops, pos, radius)
	local drop_pos = vector.new(pos)
	for _, item in pairs(drops) do
		local count = item:get_count()
		local max = item:get_stack_max()
		if count > max then
			item:set_count(max)
		end
		while count > 0 do
			if count < max then
				item:set_count(count)
			end
			rand_pos(pos, drop_pos, radius)
			local obj = core.add_item(drop_pos, item)
			if obj then
				obj:get_luaentity().collect = true
				obj:set_acceleration({x=0, y=-10, z=0})
				obj:set_velocity({x=math.random(-3, 3), y=10,
						z=math.random(-3, 3)})
			end
			count = count - max
		end
	end
end

local function add_drop(drops, item)
	item = ItemStack(item)
	local name = item:get_name()
	if loss_prob[name] ~= nil and math.random(1, loss_prob[name]) == 1 then
		return
	end

	local drop = drops[name]
	if drop == nil then
		drops[name] = item
	else
		drop:set_count(drop:get_count() + item:get_count())
	end
end

local function destroy(drops, pos, cid)
	if core.is_protected(pos, "") then
		return
	end
	local def = cid_data[cid]
	if def and def.on_blast then
		def.on_blast(vector.new(pos), 1)
		return
	end
	core.remove_node(pos)
	if def then
		local node_drops = core.get_node_drops(def.name, "")
		for _, item in ipairs(node_drops) do
			add_drop(drops, item)
		end
	end
end


local function calc_velocity(pos1, pos2, old_vel, power)
	local vel = vector.direction(pos1, pos2)
	vel = vector.normalize(vel)
	vel = vector.multiply(vel, power)

	-- Divide by distance
	local dist = vector.distance(pos1, pos2)
	dist = math.max(dist, 1)
	vel = vector.divide(vel, dist)

	-- Add old velocity
	vel = vector.add(vel, old_vel)
	return vel
end

local function entity_physics(pos, radius)
	-- Make the damage radius larger than the destruction radius
	radius = radius * 2
	local objs = core.get_objects_inside_radius(pos, radius)
	for _, obj in pairs(objs) do
		local obj_pos = obj:get_pos()
		local obj_vel = obj:get_velocity()
		local dist = math.max(1, vector.distance(pos, obj_pos))

		if obj_vel ~= nil then
			obj:set_velocity(calc_velocity(pos, obj_pos,
					obj_vel, radius * 10))
		end

		local damage = (5 / dist) * radius
		obj:set_hp(obj:get_hp() - damage)
	end
end

local function add_effects(pos, radius)
	core.add_particlespawner({
		amount = 128,
		time = 1,
		minpos = vector.subtract(pos, radius / 2),
		maxpos = vector.add(pos, radius / 2),
		minvel = {x=-20, y=-20, z=-20},
		maxvel = {x=20,  y=20,  z=20},
		minacc = vector.new(),
		maxacc = vector.new(),
		minexptime = 1,
		maxexptime = 3,
		minsize = 8,
		maxsize = 16,
		texture = "sneeker_smoke.png",
	})
end


local function explode(pos, radius)
	local pos = vector.round(pos)
	local vm = VoxelManip()
	local pr = PseudoRandom(os.time())
	local p1 = vector.subtract(pos, radius)
	local p2 = vector.add(pos, radius)
	local minp, maxp = vm:read_from_map(p1, p2)
	local a = VoxelArea:new({MinEdge = minp, MaxEdge = maxp})
	local data = vm:get_data()

	local drops = {}
	local p = {}

	local c_air = core.get_content_id("air")
	local c_tnt = nil
	if core.settings:get_bool("enable_tnt", false) then
		c_tnt = core.get_content_id("tnt:tnt")
	end

	local c_tnt_burning = core.get_content_id("tnt:tnt_burning")
	local c_gunpowder = core.get_content_id("tnt:gunpowder")
	local c_gunpowder_burning = core.get_content_id("tnt:gunpowder_burning")
	local c_boom = core.get_content_id("tnt:boom")

	for z = -radius, radius do
	for y = -radius, radius do
	local vi = a:index(pos.x + (-radius), pos.y + y, pos.z + z)
	for x = -radius, radius do
		if (x * x) + (y * y) + (z * z) <=
				(radius * radius) + pr:next(-radius, radius) then
			local cid = data[vi]
			p.x = pos.x + x
			p.y = pos.y + y
			p.z = pos.z + z
			if cid == c_tnt or cid == c_gunpowder then
				burn(p)
			elseif cid ~= c_tnt_burning and
					cid ~= c_gunpowder_burning and
					cid ~= c_air and
					cid ~= c_boom then
				destroy(drops, p, cid)
			end
		end
		vi = vi + 1
	end
	end
	end

	return drops
end

function sneeker.boom(pos, large)
	local radius = radius
	if large then
		radius = large_radius
	end
	core.sound_play("sneeker_explode", {pos=pos, gain=1.5, max_hear_distance=2*64})
	core.set_node(pos, {name="tnt:boom"})
	core.get_node_timer(pos):start(0.5)
	local drops = explode(pos, radius)
	entity_physics(pos, radius)
	eject_drops(drops, pos, radius)
	add_effects(pos, radius)
end


settings.lua
-- Settings for sneeker mod


local time_min = 60

--- How long (in seconds) sneeker remains in world after spawn.
--
--  Default: 15 minutes
--
--  @setting sneeker.lifespan
sneeker.lifetime = tonumber(core.settings:get("sneeker.lifetime") or time_min * 15)

--- Loudness of explosion.
--
--  @setting sneeker.boom_gain
sneeker.boom_gain = tonumber(core.settings:get("sneeker.boom_gain") or 1.5)

--- Determines whether or not a player must be close for spawn to occur.
--
--  @setting sneeker.spawn_require_player_nearby
sneeker.spawn_require_player_nearby = core.settings:get_bool("sneeker.spawn_require_player_nearby", true)

--- Distance in nodes a player must be for spawn to occur.
--
--  Only used if `sneeker.spawn_require_player_nearby` enabled.
--
--  @setting sneeker.spawn_player_radius
sneeker.spawn_player_radius = tonumber(core.settings:get("sneeker.spawn_player_radius") or 100)

--- If enabled, mobs not near any players will despawn.
--
--  @setting sneeker.despawn_player_far
sneeker.despawn_player_far = core.settings:get_bool("sneeker.despawn_player_far", true)

--- Distance determining if a player is near enough to prevent despawn.
--
--  Only used if `sneeker.despawn_player_far` enabled.
--
--  @setting sneeker.despawn_player_radius
sneeker.despawn_player_radius = tonumber(core.settings:get("sneeker.despawn_player_radius") or 500)

--- Sets possibility for spawn.
--
--  Inverted value (e.g. 10000 = 1/10000).
--
--  @setting sneeker.spawn_chance
sneeker.spawn_chance = tonumber(core.settings:get("sneeker.spawn_chance") or 10000)

--- Sets frequency of spawn chance.
--
--  Default 240 is equivalent to 4 minutes (60 * 4).
--
--  @setting sneeker.spawn_interval
sneeker.spawn_interval = tonumber(core.settings:get("sneeker.spawn_interval") or time_min * 4)

--- Sets the minimum light that a node must have for spawn to occur.
--
--  Default: 0
--
--  @setting sneeker.spawn_minlight
sneeker.spawn_minlight = tonumber(core.settings:get("sneeker.spawn_minlight") or 0)

--- Sets the maximum light that a node can have for spawn to occur.
--
--  Default: 4
--
--  @setting sneeker.spawn_maxlight
sneeker.spawn_maxlight = tonumber(core.settings:get("sneeker.spawn_maxlight") or 4)

--- Sets the lowest position at which sneeker can spawn.
--
--  Default: -31000
--
--  @setting sneeker.spawn_minheight
sneeker.spawn_minheight = tonumber(core.settings:get("sneeker.spawn_minheight") or -31000)

--- Sets the highest position at which sneeker can spawn.
--
--  Default: 31000
--
--  @setting sneeker.spawn_maxheight
sneeker.spawn_maxheight = tonumber(core.settings:get("sneeker.spawn_maxheight") or 31000)

--- Limits the number of entities that can spawn per mapblock (16x16x16).
--
--  Default: 1
sneeker.spawn_mapblock_limit = tonumber(core.settings:get("sneeker.spawn_mapblock_limit") or 1)
